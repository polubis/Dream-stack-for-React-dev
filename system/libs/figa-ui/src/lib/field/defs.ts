import type { ReactNode } from 'react';
import type { BoxProps } from '../box';

interface FieldProps extends BoxProps {
  label?: ReactNode;
  hint?: ReactNode;
  error?: ReactNode;
}

export type { FieldProps };
