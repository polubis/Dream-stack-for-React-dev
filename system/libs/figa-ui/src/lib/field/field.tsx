import { Box } from '../box';
import { Font } from '../font';
import type { FieldProps } from './defs';

import c from 'classnames';

const Field = ({
  className,
  hint,
  children,
  error,
  label,
  ...props
}: FieldProps) => {
  return (
    <Box className={c('field', className, { invalid: !!error })} {...props}>
      {label && (
        <Font variant="b2" className="field-label">
          {label}
        </Font>
      )}
      {children}
      {(!!error || !!hint) && (
        <Font italic variant="b3" className="field-hint">
          {!error && hint}
          {error}
        </Font>
      )}
    </Box>
  );
};

export { Field };
