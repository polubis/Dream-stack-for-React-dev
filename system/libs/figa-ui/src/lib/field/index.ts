export * from './field';
export * from './defs';
