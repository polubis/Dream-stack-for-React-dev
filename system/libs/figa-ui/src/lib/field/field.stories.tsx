import type { Story, Meta } from '@storybook/react';

import { Field } from './field';
import { Box } from '../box';
import { Input } from '../input';
import { Font } from '../font';

export default {
  component: Field,
  title: 'Field',
} as Meta;

const Template: Story = () => {
  return (
    <Box
      padding={[300, 300, 300, 300]}
      spacing={[100, 300, 100, 300, 100, 300, 100, 300, 100, 300]}
    >
      <Font variant="h5">Field without hint</Font>
      <Field>
        <Input placeholder="Type some value..." />
      </Field>
      <Font variant="h5">Field with hint</Font>
      <Field hint="Some description...">
        <Input placeholder="Type some value..." />
      </Field>
      <Font variant="h5">Field with error</Font>
      <Field hint="Some description..." error="Some error...">
        <Input placeholder="Type some value..." />
      </Field>
      <Font variant="h5">Field with error and label</Font>
      <Field
        maxWidth='300px'
        label="Example of label ads adda ssada dsadsa dsadsad asd sad saasdasdasdad "
        hint="Some description..."
        error="Some error..."
      >
        <Input placeholder="Type some value..." />
      </Field>
    </Box>
  );
};

export const Default = Template.bind({});
Default.args = {};
