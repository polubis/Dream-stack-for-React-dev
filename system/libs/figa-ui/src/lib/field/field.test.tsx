import { fireEvent, render, screen } from '@testing-library/react';
import { Field } from './field';

describe('Field can be used when: ', () => {
  it('[FRAGILE] assigns classes if error passed', () => {
    const { asFragment } = render(
      <Field className="my-class" error="My error" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('[FRAGILE] assigns classes if no error passed', () => {
    const { asFragment } = render(
      <Field className="my-class" hint="My content" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('allows to pass any other native property', () => {
    const spy = jest.fn();

    const { asFragment } = render(
      <Field
        hint="My content"
        maxWidth="300px"
        minWidth="200px"
        spacing={[200, 200, 200, 200]}
        onClick={spy}
      />
    );

    fireEvent.click(screen.getByText(/My content/));

    expect(asFragment()).toMatchSnapshot();
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
