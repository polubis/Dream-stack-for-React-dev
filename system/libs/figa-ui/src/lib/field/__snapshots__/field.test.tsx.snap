// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field can be used when:  [FRAGILE] assigns classes if error passed 1`] = `
<DocumentFragment>
  <div
    class="box field my-class invalid column empty"
  >
    <p
      class="font b3 default field-hint italic"
    >
      My error
    </p>
  </div>
</DocumentFragment>
`;

exports[`Field can be used when:  [FRAGILE] assigns classes if no error passed 1`] = `
<DocumentFragment>
  <div
    class="box field my-class column empty"
  >
    <p
      class="font b3 default field-hint italic"
    >
      My content
    </p>
  </div>
</DocumentFragment>
`;

exports[`Field can be used when:  allows to pass any other native property 1`] = `
<DocumentFragment>
  <div
    class="box field column empty"
    style="max-width: 300px; min-width: 200px;"
  >
    <div
      class="box-item-wrapper"
      style="margin-bottom: 16px;"
    >
      <p
        class="font b3 default field-hint italic"
      >
        My content
      </p>
    </div>
  </div>
</DocumentFragment>
`;
